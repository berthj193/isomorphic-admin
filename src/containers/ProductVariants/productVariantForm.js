import React from 'react';
import PropTypes from 'prop-types';
import { Field } from 'redux-form';
import { Icon } from 'antd';
import Form from '../../components/uielements/form';
import ButtonList from '../../components/uielements/buttonList';
import OptionsInput from '../../components/uielements/optionsInput';
import Input from '../../components/uielements/input';
import makeField from '../../utils/makeField';

import { required, fieldListRequired } from '../../helpers/formValidation';

export default class ProductVariantForm extends React.Component {
  static propTypes = {
    productId: PropTypes.string,
    onDelete: PropTypes.func,
    handleSubmit: PropTypes.func,
    dirty: PropTypes.bool,
    pristine: PropTypes.bool,
  };

  render() {
    const { productId } = this.props;
    const Buttons = makeField(() => (
      <ButtonList
        buttons={[
          { onClick: this.props.handleSubmit,
            type: 'primary',
            key: `productVariantSave${productId}`,
            title: (
              <React.Fragment>
                <Icon type="check" />
                <span />
                Save
              </React.Fragment>),
          },
          { onClick: this.props.onDelete,
            type: 'danger',
            key: `productVariantDelete${productId}`,
            ghost: true,
            title: (
              <React.Fragment>
                <Icon type="delete" />
                <span />
                Delete
              </React.Fragment>),
          },
        ]} />
    ), { withoutLabel: true });
    return (
      <Form onSubmit={this.props.handleSubmit}>
        <Field
          label="Default SKU"
          name="default_sku"
          component={Input}
          validate={[required]}
          required
        />
        <Field
          name="default_options"
          component={OptionsInput}
          validate={[fieldListRequired]}
          required
        />
        <Buttons />
      </Form>
    );
  }
}
