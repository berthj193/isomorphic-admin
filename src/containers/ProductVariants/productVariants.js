import React from 'react';
import PropTypes from 'prop-types';

import productQueries from '../../graphql/products';
import productVariantQueries from '../../graphql/productVariants';

import withReduxForm from '../../components/formWrapper/withReduxForm';
import withStoresManager from '../../components/apolloProviders/withStoresManager';
import ProductVariantForm from './productVariantForm';
import LayoutContentWrapper from '../../components/utility/layoutWrapper';
import LayoutContent from '../../components/utility/layoutContent';
import { FormDivider } from '../../components/uielements/form';
import Button from '../../components/uielements/button';
import ButtonList from '../../components/uielements/buttonList';
import makeField from '../../utils/makeField';
import _isEqual from 'lodash/isEqual';

export class ProductVariants extends React.Component {
  static propTypes = {
    product_variants: PropTypes.arrayOf(PropTypes.shape({
      default_options: PropTypes.object,
      default_sku: PropTypes.string,
      id: PropTypes.string,
    })),
    match: PropTypes.shape({
      params: PropTypes.shape({
        id: PropTypes.string,
      }),
    }),
    mutate: PropTypes.func,
    products: PropTypes.arrayOf(PropTypes.shape({
      name: PropTypes.string,
    })),
  };

  static defaultProps = {
    product_variants: [],
  };

  state = {
    productVariants: [],
    forms: [],
  };

  componentDidMount() {
    this.productVariantInitialValues({});
  }

  componentDidUpdate(prevProps) {
    this.productVariantInitialValues(prevProps);
  }

  productVariantInitialValues = prevProps => {
    const { product_variants } = this.props;
    if (!_isEqual(product_variants, prevProps.product_variants) && product_variants.length) {
      const productVariants = [];
      const forms = [];
      product_variants.forEach(productVariant => {
        forms.push(this.addNewForm(forms.length));

        productVariants.push({
          ...productVariant,
          default_options: productVariant.default_options
            && Object.entries(productVariant.default_options).map(option => ({
              key: option[0],
              value: option[1],
            })),
        });
      });

      this.setState({
        productVariants,
        forms,
      });
    }
  };

  get productId() {
    return this.props.match.params.id;
  }

  addNewForm = formId =>
    withReduxForm({
      form: `productVariantForm${this.productId}-${formId}`,
    })(ProductVariantForm);

  addProductVariant = () => this.setState(state => ({
    ...state,
    productVariants: [
      ...state.productVariants,
      {},
    ],
    forms: [
      ...state.forms,
      this.addNewForm(state.forms.length),
    ],
  }));

  handleSave = values => {
    let mutation = productVariantQueries.updateProductVariant;
    const default_values = values.default_options.reduce((prevValue, { key, value }) => ({
      ...prevValue,
      [key]: value,
    }), {});
    const variables = {
      ...values,
      default_options: JSON.stringify(default_values),
    };

    if (!values.id) {
      mutation = productVariantQueries.createProductVariant;
      variables.product_id = this.productId;
    }

    this.props.mutate({
      mutation,
      variables,
      refetchQueries: [
        { query: productVariantQueries.getList },
      ],
      messages: ['Product variant modified successfully'],
    });
  };

  removeProductVariant = index => {
    this.setState(state => {
      state.productVariants[index] = null;
      state.forms[index] = null;
      return state;
    });
  };

  handleDelete = index => () => {
    const { id } = this.state.productVariants[index];
    if (id) {
      this.props.mutate({
        mutation: productVariantQueries.destroyProductVariant,
        variables: {
          id,
        },
        refetchQueries: [
          { query: productVariantQueries.getList },
        ],
      }).then(() => {
        this.removeProductVariant(index);
      });
    } else {
      this.removeProductVariant(index);
    }
  };

  renderProductVariantForms = id =>
    this.state.productVariants.map((productVariant, index) => {
      if (!productVariant) {
        return;
      }
      const Form = this.state.forms[index];
      return (
        <React.Fragment key={productVariant.id || `freshProductVariant${index}`}>
          <Form
            productId={id}
            initialValues={productVariant}
            onDelete={this.handleDelete(index)}
            onSubmit={this.handleSave}
          />
          <FormDivider />
        </React.Fragment>
      );
    });

  render() {
    const { products = [{}] } = this.props;

    const product = products.find(({ id }) => id === this.productId) || {};
    const ProductName = makeField(() => product.default_name || '');

    const AddProductVariantButton = makeField(
      () => <Button onClick={this.addProductVariant}>Add product variant</Button>,
      { withoutLabel: true }
    );

    return (
      <LayoutContentWrapper style={{ minHeight: '100vh' }}>
        <LayoutContent>
          <h1>Product variants</h1>
          <ButtonList buttons={[{ link: `/dashboard/products`, title: 'Go back' }]} />
          <ProductName label="Product name" />
          <FormDivider />
          { this.renderProductVariantForms(this.productId) }
          <AddProductVariantButton />
        </LayoutContent>
      </LayoutContentWrapper>
    );
  }
}

export default withStoresManager(ProductVariants, [
  {
    query: productVariantQueries.getList,
  },
  {
    query: productQueries.productsAutocomplete,
  },
]);
