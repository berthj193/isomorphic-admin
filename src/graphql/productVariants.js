import gql from 'graphql-tag';

export const getListQuery = `
query {
  product_variants {
    id
    default_sku
    default_options
    product {
      id
      default_master_sku
    }
    store_product_variants {
      id
      sku
    }
  }
}`;

export const createProductVariantQuery = `
mutation createProductVariant(
  $default_sku:String!, $default_options:JsonType!,
  $product_id:ID!
) {
  createProductVariant(
    default_sku: $default_sku
    default_options: $default_options
    product_id: $product_id
  ) {
    id
  }
}`;

export const productVariantAutocompleteQuery = `
query {
  product_variants {
    id
    default_sku
  }
}`;

export const updateProductVariantQuery = `
mutation updateProductVariant(
  $default_sku:String!, $default_options:JsonType!,
  $id:ID!
) {
  updateProductVariant(
    default_sku: $default_sku
    default_options: $default_options
    id: $id
  ) {
    id
  }
}`;

export const destroyProductVariantQuery = `
mutation destroyProductVariant (
  $id: ID!
) {
  destroyProductVariant (
    id: $id
  )
}`;

export default {
  getList: gql`${getListQuery}`,
  createProductVariant: gql`${createProductVariantQuery}`,
  updateProductVariant: gql`${updateProductVariantQuery}`,
  destroyProductVariant: gql`${destroyProductVariantQuery}`,
  productVariantAutocomplete: gql`${productVariantAutocompleteQuery}`,
};
