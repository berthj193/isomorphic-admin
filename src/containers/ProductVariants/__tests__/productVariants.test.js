import React from 'react';
import { ProductVariants } from '../productVariants';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

describe('<productVariants />', () => {
  const defaultProps = {
    match: {
      params: {
        id: '1',
      },
    },
  };
  const getComponent = (props = {}) =>
    shallow(<ProductVariants {...defaultProps} {...props} />);

  it('should render properly', () => {
    expect(toJson(getComponent())).toMatchSnapshot();
  });

  it('should add product variant from props to state in componentDidMount() lifecycle method', () => {
    const wrapper = getComponent({
      product_variants: [{
        options: { foo: 'bar' },
      }],
    });
    expect(wrapper.instance().state.productVariants.length).toBe(1);
  });

  it('productId should return URL param', () => {
    const wrapper = getComponent();
    expect(wrapper.instance().productId).toBe('1');
  });

  it('addProductVariant() should add product variant to state', () => {
    const wrapper = getComponent();
    const instance = wrapper.instance();
    instance.addProductVariant();
    expect(instance.state.productVariants.length).toBe(1);
  });

  it('handleSave() should execute mutate function', () => {
    const wrapper = getComponent({
      mutate: jest.fn().mockReturnValue(Promise.resolve('data')),
    });
    const instance = wrapper.instance();
    instance.handleSave({
      default_options: [{ key: 'foo', value: 'bar' }],
    });
    expect(instance.props.mutate).toHaveBeenCalledTimes(1);
  });

  it('handleDelete() should delete product variant from state', () => {
    const wrapper = getComponent();
    const instance = wrapper.instance();
    instance.addProductVariant();
    instance.handleDelete(0)();
    expect(instance.state.productVariants[0]).toBe(null);
  });

  it('handleDelete() should execute mutate function if product variant ID is provided', () => {
    const wrapper = getComponent({
      mutate: jest.fn().mockReturnValue(Promise.resolve('data')),
      product_variants: [{ id: '1' }],
    });
    const instance = wrapper.instance();
    instance.handleDelete(0)();
    expect(instance.props.mutate).toHaveBeenCalledTimes(1);
  });

  it('renderProductVariantForms() should render product variant forms based on state data', () => {
    const wrapper = getComponent();
    const instance = wrapper.instance();
    instance.setState({
      productVariants: [{
        id: '1',
        default_sku: 'foo',
      }],
      forms: [
        require('../productVariantForm').default,
      ],
    });
    expect(instance.renderProductVariantForms('1')).toMatchSnapshot();
  });
});
