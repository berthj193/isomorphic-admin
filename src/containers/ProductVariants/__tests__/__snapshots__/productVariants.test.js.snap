// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<productVariants /> renderProductVariantForms() should render product variant forms based on state data 1`] = `
Array [
  <Unknown>
    <ProductVariantForm
        initialValues={
            Object {
                "default_sku": "foo",
                "id": "1",
              }
        }
        onDelete={[Function]}
        onSubmit={[Function]}
        productId="1"
    />
    <FormDivider />
</Unknown>,
]
`;

exports[`<productVariants /> should render properly 1`] = `
<LayoutWrapper
  style={
    Object {
      "minHeight": "100vh",
    }
  }
>
  <LayoutContent>
    <h1>
      Product variants
    </h1>
    <ButtonList
      buttons={
        Array [
          Object {
            "link": "/dashboard/products",
            "title": "Go back",
          },
        ]
      }
    />
    <CustomComponent
      label="Product name"
    />
    <FormDivider />
    <CustomComponent />
  </LayoutContent>
</LayoutWrapper>
`;
